File: server/src/Vitals.Domain/Fact.cs
namespace Vitals.Domain;

/// <summary>What kind of thing a fact records. The number is the encoding's; add at the end, never renumber.</summary>
public enum FactKind : byte
{
    Registration = 0,
    Vitals = 1,
    Immunisation = 2,
    AncVisit = 3,
    StockMovement = 4,
    Note = 5,
    Supersession = 6,
    Access = 7,
}

/// <summary>One immutable observation. Equality is by id; order is by stamp then id.</summary>
public sealed class Fact : IComparable<Fact>, IEquatable<Fact>
{
    public Fact(byte[] id, byte[] patient, FactKind kind, Stamp stamp, string author, byte[] payload, byte[]? supersedes)
    {
        Id = id; Patient = patient; Kind = kind; Stamp = stamp; Author = author; Payload = payload; Supersedes = supersedes;
        Key = Convert.ToHexStringLower(id);
    }

    public byte[] Id { get; }
    public byte[] Patient { get; }
    public FactKind Kind { get; }
    public Stamp Stamp { get; }
    public string Author { get; }
    public byte[] Payload { get; }
    public byte[]? Supersedes { get; }
    public string Key { get; }

    public int CompareTo(Fact? other)
    {
        if (other is null) return 1;
        var c = Stamp.CompareTo(other.Stamp);
        return c != 0 ? c : Bytes.Compare(Id, other.Id);
    }

    public bool Equals(Fact? other) => other is not null && Key == other.Key;
    public override bool Equals(object? obj) => Equals(obj as Fact);
    public override int GetHashCode() => Key.GetHashCode(StringComparison.Ordinal);
}

public static class Bytes
{
    public static int Compare(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b)
    {
        var n = Math.Min(a.Length, b.Length);
        for (var i = 0; i < n; i++)
        {
            if (a[i] != b[i]) return a[i].CompareTo(b[i]);
        }
        return a.Length.CompareTo(b.Length);
    }

    public static bool Same(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b) => Compare(a, b) == 0;
}

File: packages/vitals_domain/lib/src/fact.dart
import 'clock.dart';

/// What kind of thing a fact records. The number is the encoding's; add at the
/// end, never renumber (ADR-0003: the encoding never changes).
enum FactKind {
  registration(0),
  vitals(1),
  immunisation(2),
  ancVisit(3),
  stockMovement(4),
  note(5),
  supersession(6),
  access(7);

  const FactKind(this.code);
  final int code;

  static FactKind fromCode(int code) =>
      FactKind.values.firstWhere((k) => k.code == code);
}

/// One immutable observation. Who, when, on which device, about whom, what.
/// A fact never changes; a correction is a [FactKind.supersession] fact naming
/// the id it replaces, and both survive.
final class Fact implements Comparable<Fact> {
  const Fact({
    required this.id,
    required this.patient,
    required this.kind,
    required this.stamp,
    required this.author,
    required this.payload,
    this.supersedes,
  });

  /// 32 bytes, chosen by the app from randomness the domain never sees.
  final List<int> id;

  /// The patient's global id, 16 bytes.
  final List<int> patient;
  final FactKind kind;
  final Stamp stamp;

  /// The staff member, by their attribution id. Every write is attributed.
  final String author;

  /// The observation itself, already encoded by its kind's own canonical
  /// encoder. Opaque here; the record does not read it, only keeps it.
  final List<int> payload;

  /// For a supersession: the id of the fact it corrects.
  final List<int>? supersedes;

  String get key => _hex(id);

  @override
  int compareTo(Fact other) {
    final c = stamp.compareTo(other.stamp);
    return c != 0 ? c : _compareBytes(id, other.id);
  }

  @override
  bool operator ==(Object other) => other is Fact && _sameBytes(other.id, id);

  @override
  int get hashCode => Object.hashAll(id);
}

int _compareBytes(List<int> a, List<int> b) {
  final n = a.length < b.length ? a.length : b.length;
  for (var i = 0; i < n; i++) {
    if (a[i] != b[i]) return a[i].compareTo(b[i]);
  }
  return a.length.compareTo(b.length);
}

bool _sameBytes(List<int> a, List<int> b) => _compareBytes(a, b) == 0;

String _hex(List<int> bytes) {
  const digits = '0123456789abcdef';
  final out = StringBuffer();
  for (final b in bytes) {
    out.write(digits[b >> 4]);
    out.write(digits[b & 0xf]);
  }
  return out.toString();
}
